contracts/delegation/dao-vote-delegation/src/contract.rs

```diff
@@ -0,0 +1,757 @@
+            return Err(ContractError::InvalidDelegationValidityBlocks {
+                provided: delegation_validity_blocks,
+                min: 2,
+            });
+        }
+    }
+
+    CONFIG.save(
+        deps.storage,
+        &Config {
+            delegation_validity_blocks: msg.delegation_validity_blocks,
+            max_delegations: msg.max_delegations.unwrap_or(DEFAULT_MAX_DELEGATIONS),
+        },
+    )?;
+    VP_CAP_PERCENT.save(deps.storage, &msg.vp_cap_percent, env.block.height)?;
+
+    // initialize voting power changed hook callers
+    if let Some(vp_hook_callers) = msg.vp_hook_callers {
+        for caller in vp_hook_callers {
+            VOTING_POWER_HOOK_CALLERS.save(deps.storage, deps.api.addr_validate(&caller)?, &())?;
+        }
+    }
+
+    // sync proposal modules with no limit if not disabled. this should succeed
+    // for most DAOs as the query will not run out of gas with only a few
+    // proposal modules.
+    if !msg.no_sync_proposal_modules.unwrap_or(false) {
+        execute_sync_proposal_modules(deps, None, None)?;
+    }
+
+    Ok(Response::new().add_attribute("dao", dao))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Register {} => execute_register(deps, env, info),
+        ExecuteMsg::Unregister {} => execute_unregister(deps, env, info),
+        ExecuteMsg::Delegate { delegate, percent } => {
+            execute_delegate(deps, env, info, delegate, percent)
+        }
+        ExecuteMsg::Undelegate { delegate } => execute_undelegate(deps, env, info, delegate),
+        ExecuteMsg::UpdateVotingPowerHookCallers { add, remove } => {
+            execute_update_voting_power_hook_callers(deps, info, add, remove)
+        }
+        ExecuteMsg::SyncProposalModules { start_after, limit } => {
+            execute_sync_proposal_modules(deps, start_after, limit)
+        }
+        ExecuteMsg::UpdateConfig {
+            vp_cap_percent,
+            delegation_validity_blocks,
+            max_delegations,
+        } => execute_update_config(
+            deps,
+            env,
+            info,
+            vp_cap_percent,
+            delegation_validity_blocks,
+            max_delegations,
+        ),
+        ExecuteMsg::StakeChangeHook(msg) => execute_stake_changed(deps, env, info, msg),
+        ExecuteMsg::NftStakeChangeHook(msg) => execute_nft_stake_changed(deps, env, info, msg),
+        ExecuteMsg::MemberChangedHook(msg) => execute_membership_changed(deps, env, info, msg),
+        ExecuteMsg::VoteHook(vote_hook) => execute_vote_hook(deps, env, info, vote_hook),
+    }
+}
+
+fn execute_register(deps: DepsMut, env: Env, info: MessageInfo) -> Result<Response, ContractError> {
+    nonpayable(&info)?;
+    ensure_setup(deps.as_ref())?;
+
+    let delegate = info.sender;
```

Comment:
wyt if we move the `nonpayable` call into the execute entry point handler before the match statement?
that way we could avoid passing `MessageInfo` just in order to do `let delegate/delegator = info.sender` inside the functions. on the other hand, that would mean that the hooks incur a redundant check.

Reply:
ah yes great idea, since none of the contract takes funds. i'll push that here